lte/gateway/c/session_manager/SessionState.cpp

```diff
@@ -458,11 +463,87 @@ bool SessionState::apply_update_criteria(SessionStateUpdateCriteria& uc) {
   return true;
 }
 
+RuleStats SessionState::get_rule_delta(
+    const std::string& rule_id, uint64_t rule_version, uint64_t used_tx,
+    uint64_t used_rx, uint64_t dropped_tx, uint64_t dropped_rx,
+    SessionStateUpdateCriteria& uc) {
+  RuleStats ret = {0};
+
+  auto it = policy_version_and_stats_.find(rule_id);
+  if (it == policy_version_and_stats_.end()) {
+    MLOG(MERROR) << "Reported rule not found, ignoring";
+    return ret;
+  }
+
+  // Only accept rule reports for current_version or last_reported_version
+  // ignore other reports as they shoudn't be sent
+  auto last_reported_version = it->second.last_reported_version;
+  if (rule_version > it->second.current_version) {
+    MLOG(MERROR) << "Reported version higher than tracked one, ignoring";
+    return ret;
+  }
+
+  if (rule_version < last_reported_version) {
+    MLOG(MERROR) << "Reported rule version too old, ignoring";
+    return ret;
+  }
+
+  auto last_tracked =
+      policy_version_and_stats_[rule_id].stats_map.find(last_reported_version);
+
+  RuleStats prev_usage = last_tracked->second;
+
+  if (rule_version == last_reported_version) {
```

Comment:
Per what I see rule version is reported by Pipelined. If we report the same version that means we get the delta between pipelined current report and whatever we have stored in sessiond

But if version has changed, we reset the value? Aren't we going to lose data if that hasn't been reported yet? Why change on version of the rule (being the same rule id) should change its counter?